operator/internal/controller/auto/update_controller.go

```diff
@@ -504,3 +521,40 @@ func (r *UpdateReconciler) mapWorkspaceToUpdate(ctx context.Context, obj client.
 	}
 	return requests
 }
+
+// outputsToSecret returns a Secret object whose keys are stack output names
+// and values are JSON-encoded bytes. An annotation is recorded with all secret
+// outputs overwritten with "[secret]"; this annotation is consumed by the
+// Stack API and recorded on the stack's status.
+func outputsToSecret(owner *autov1alpha1.Update, outputs map[string]*agentpb.OutputValue) (*corev1.Secret, error) {
+	s := &corev1.Secret{Immutable: ptr.To(true), Data: map[string][]byte{}}
+	s.SetName(owner.Name + "-stack-outputs")
+	s.SetNamespace(owner.Namespace)
+	s.SetOwnerReferences([]metav1.OwnerReference{{
+		APIVersion: owner.APIVersion,
+		Kind:       owner.Kind,
+		Name:       owner.Name,
+		UID:        owner.UID,
+	}})
+
+	scrubbed := make(map[string]any, len(outputs))
+	for k, v := range outputs {
+		// v.Value is already JSON-encoded bytes,
+		s.Data[k] = v.Value
+		if v.Secret {
+			scrubbed[k] = "[secret]"
+		} else {
+			scrubbed[k] = json.RawMessage(v.Value)
+		}
+	}
+
+	annotation, err := json.Marshal(scrubbed)
+	if err != nil {
+		return nil, fmt.Errorf("marshaling scrubbed output: %w", err)
+	}
+	s.SetAnnotations(map[string]string{
+		"scrubbed": string(annotation),
```

Comment:
Shall we use a namepaced annotation name? e.g. `pulumi.com/scrubbed`.

Reply:
Changed to pulumi.com/mask